### ISSUE TITLE
Coveralls reporting 0% coverage with Slather/Travis integration

### ISSUE THREAD

**Issue by smileyborg** [venmo/slather#69]
So this is the first time I've used Slather + Coveralls, and I'm running into an issue as described here: https://github.com/intuit/LocationManager/pull/37#issuecomment-94590812

Running Slather locally on the Xcode project works fine:

```
$ slather coverage -s LocationManagerExample/LocationManagerExample.xcodeproj/
Slathering...
Source/INTULocationManager.m: 180 of 240 lines (75.00%)
Source/INTULocationRequest.m: 60 of 73 lines (82.19%)
Test Coverage: 76.68%
Slathered
```

However, something is failing when this runs on Travis, and as a result Coveralls [shows 0% coverage](https://coveralls.io/r/intuit/LocationManager) (although Coveralls does show the correct source files). Here's the output from a Travis build: https://travis-ci.org/intuit/LocationManager/builds/59321046

Any idea if this a Slather or Coveralls issue and what I need to do to get this working?

**Comment by marklarr** [comment]
Hey, thanks for filing an issue. I've been hearing a couple different of people have had this issue today, so I'm going to dig into it tonight a little bit!

**Comment by lazerwalker** [comment]
+1. I set up two projects with Slather + Coveralls yesterday; one is encountering this issue while the other is fine. So while I have no idea what might be causing it, it's not a universal thing.

**Comment by marklarr** [comment]
@kylef, I want to say that I _vaguely_ remember something about the [Cocoapods post-install hook](https://github.com/venmo/slather/blob/master/lib/cocoapods_plugin.rb) not consistently working with Travis in the beginning... or at least having some sort of issue?

 The reason that I'd believe that it might hae something to do with that plugin is because slather is setup to blow-up if it [cannot find any coverage files](https://github.com/venmo/slather/blob/d3e18c15c1b86c1fc08c6294c3054bdb4fb1a348/lib/slather/project.rb#L49) I _think_ that each project does generate `gcno` files for every source file, though, so if a target is generating `gcno` files that doesn't contain the test suite for said files, then they'll show up with 0 coverage down the boards. Usually slather just circumvents this by finding and using the `gcno` files with the highest amount of coverage, assuming those to be correct. But maybe those correct ones were never created because the Pods project didn't get configured for `instrument program flow` and `generate coverage files` correctly?

I feel that there's a 50/50 chance that paragraph even made any sense :disappointed:

**Comment by lazerwalker** [comment]
The intuit/LocationManager project @smileyborg mentioned isn't using the CP plugin, though. The lib files are included directly in the example project, and its Podfile only exists to add Specta/Expecta.

Poking through the Slather code, it doesn't look like it has a `--verbose` flag or anything like that? Getting a little more insight into what's actually going on when it runs on Travis could be enlightening.

**Comment by smileyborg** [comment]
+1 for a `--verbose` flag; I was wondering if it's possible to use the `-s` option to get command line output in combination with the Coveralls integration.

**Comment by marklarr** [comment]
Yea, I've been thinking about the verbose flag this morning. I was thinking the following things. Let me know what you all think --also, @kylef @ayanonagon @dasmer @eliperkins 

(for verbose flag)
* Log when we've successfully configured pods for slather
* Log each GCNO file that is ignored (ones for files that aren't your source files, and also duplicate ones.)
* Log each GCNO file that is processed, along with its 
  * created_at timestamp
  * modified_at timestamp
  * Contents? Seems too noisy... but it'd be nice to at least see a couple.
  * Add some sort of `#inspect` method to `Slather::CoverageFile` class, and print out this as well.
* Add coverage_service specific logging to each of the services. 
  * For example, for coveralls, print out the JSON blob that it'll send to coveralls, and the response from coveralls after it has been posted.

(not related to verbose flag)
* Allow `.slather.yml` files to use multiple coverage services with `coverage_services:`, and allow the flags to the command line to stack, ie `-s` and `-x` can be used together for cobertura and terminal output.

Is there anything else I should add that would be helpful?

**Comment by dasmer** [comment]
@marklarr I agree that contents might be to noisy so maybe leave that out, but everything else would be really helpful

**Comment by dasmer** [comment]
@smileyborg You can use `-s` option in your Travis file, but you will still get 0% coverage.

See:
https://github.com/AlexDenisov/FrameAccessor/pull/17
![image](https://cloud.githubusercontent.com/assets/265892/7281437/a3bbd84c-e8f7-11e4-9695-1c98cfa6504d.png)

**Comment by smileyborg** [comment]
@dasmer That's helpful. So that means that the problem is with Slather picking up the coverage when running on Travis, (probably) not anything with the communication from Slather to Coveralls.

**Comment by smileyborg** [comment]
@marklarr May I suggest opening a new Issue to discuss adding a `--verbose` option, and associating commits to that one? Going to sidetrack this one.

**Comment by ayanonagon** [comment]
https://github.com/venmo/slather/issues/71

**Comment by marklarr** [comment]
Thanks everyone

**Comment by smileyborg** [comment]
OK, as far as the original issue here, I had a theory about what might be going on, and I was right.

This was the original `.travis.yml` that was being used:
```
sudo: false
language: objective-c
cache: cocoapods
before_install:
  - export LANG=en_US.UTF-8
  - gem install cocoapods xcpretty slather --no-ri --no-rdoc
  - pod install --project-directory=LocationManagerExample
  
script:
  - set -o pipefail && xcodebuild test -workspace LocationManagerExample/LocationManagerExample.xcworkspace -scheme LocationManagerExample -sdk iphonesimulator ONLY_ACTIVE_ARCH=NO | xcpretty -c

after_success: slather
```

Note that the tests were being run manually using the `script` key, instead of letting Travis do its default execution using `xctool`.

By removing this `script` key and changing the `.travis.yml` to:
```
sudo: false
language: objective-c
cache: cocoapods
xcode_workspace: LocationManagerExample/LocationManagerExample.xcworkspace
xcode_scheme: LocationManagerExample
xcode_sdk: iphonesimulator
podfile: LocationManagerExample/Podfile

before_install:
  - gem install cocoapods --no-ri --no-rdoc
  - gem install slather --no-ri --no-rdoc

after_success: slather
```

Travis is now in charge of invoking `xctool` to run the tests, and Slather is now able to correctly get the coverage data and send it over to Coveralls: https://coveralls.io/r/intuit/LocationManager

So, not sure if you'd consider this a bug -- is Slather supposed to support the manual mechanism of test execution that was happening originally?

**Comment by marklarr** [comment]
`slather` should work wtih vanilla `xcodebuild`, or even running the tests via Xcode. The only consideration that should need to be taken is setting the `build_directory` in your `.slather.yml`, as mentioned [here](https://github.com/venmo/slather#custom-build-directory). [Code](https://github.com/venmo/slather/blob/master/lib/slather/project.rb#L38). 

It's pretty odd that it changes anything, really. I wonder what `xcodebuild` command travis ends up calling when you had it figure it out on it's own? Does it show it in the travis logs?

**Comment by dasmer** [comment]
@smileyborg 
https://github.com/lazerwalker/MTLParseAdapter also uses `script` with `xcodebuild` and it just works

**Comment by marklarr** [comment]
Are we entirely sure that this failure is deterministic?

@dasmer I know that the guys at Travis [are switching their OSX providers](http://blog.travis-ci.com/2015-04-09-this-week-in-travis-ci/), maybe they are slowly rolling that out, and `slather` doesn't play well with the setup on the new machines?@joshk @meatballhatt

**Comment by smileyborg** [comment]
@marklarr Yes, Travis outputs the build commands it's using, just go to this recent build (which successfully slathered coverage data to Coveralls) and search the log for `xctool`: https://travis-ci.org/intuit/LocationManager/builds/59640143

**Comment by smileyborg** [comment]
@marklarr And I'm not 100% certain this was deterministic, but a bunch of builds failed to slather correctly before this change, and as soon as I made the change it's working consistently...

**Comment by marklarr** [comment]
@dasmer @lazerwalker, I'd say that we should try what @smileyborg has done on IntentKit and FrameAccessor and see what happens. It's not an ideal solution, but it will at least give us more of an idea of what is causing the issue.

**Comment by lazerwalker** [comment]
Yeah, I'm game to give it a shot.

**Comment by lazerwalker** [comment]
That fixed Coveralls for IntentKit, but any of you know how to run things through `xcpretty` without using `script`? The ugly output of https://travis-ci.org/intentkit/IntentKit/builds/59646820 is seriously bumming me out.

**Comment by dasmer** [comment]
This fixed FrameAccessor to :dancer:

**Comment by smileyborg** [comment]
If I had to speculate, I would say it could be that the build output directory is different when using `script` or something (perhaps Travis recently changed it?) and Slather is unable to find the coverage files where it expects to find them.

**Comment by marklarr** [comment]
@smileyborg I agree with your speculation, but the _one_ weird thing about that is that if Slather doesn't find any coverage files, it [raises an exception](https://github.com/venmo/slather/blob/master/lib/slather/project.rb#L49). So, it's defintely finding `gcno` files, but they're just reporting no coverage, which usually would happen in the case of the build settings for `GCC_INSTRUMENT_PROGRAM_FLOW_ARCS` and `GCC_GENERATE_TEST_COVERAGE_FILES` not being enabled.

**Comment by marklarr** [comment]
@lazerwalker I think we're SOL on using `xcpretty` without `script`. At least we've narrowed it down to `script` being the issue now, though, and maybe we can figure this out once and for all. I haven't had a lot of time to look at this these last couple days, but the first chance I get to work on anything OS it'll be on this issue.

**Comment by mkalmes** [comment]
Hey folks! :wave: 

I'm expiriencing the same behavior, as described earlier. I've tried to setup my project (removing pod linting and ditching xcpretty) to mimic the setup of [IntentKit](https://github.com/intentkit/IntentKit).

Travis will happily build the project but slather will now [raise an exception](https://travis-ci.org/trustedshops/trustbadge_iOS/builds/73021829) that it can't find the coverage files. Previously I've provided the necessary build options directly [in the script](https://github.com/trustedshops/trustbadge_iOS/commit/b12e083633192708ce827c2bd26e615b051e2738). [Right now](https://github.com/trustedshops/trustbadge_iOS/commit/484e5db9befbe803fd46acf5e5bc8588edcd390b) the creation of the coverage file is added to the project file.

Any ideas what I can try next to get the coverage report running again?

**Comment by timbodeit** [comment]
I am currently trying to integrate slather with coveralls and travis into [nui](https://github.com/tombenner/nui).

We are having a rather large build matrix on travis for that project. Code coverage is generated fine except for a single build with xcode6.1 (travis default) run on a 7.0.3 Simulator.

This bug may be specific to individual iOS versions.

See:  
[Coveralls](https://coveralls.io/builds/4198656)  
[Travis Build](https://travis-ci.org/timbodeit/nui/builds/91933844)

Xcode versions for the matrix and output of `slather coverage -s` included in travis build

**Comment by yas375** [comment]
I also wasn't getting any coverage on coveralls.io (https://coveralls.io/builds/6001539). On another run it showed list of files, but was missing 1/3 of files: https://coveralls.io/builds/6001574. I was using slather 2.1.0 at that time.

Running `slather coverage` locally was producing correct results. Locally I've debugged slather gem and it also seemed to send correct data (https://gist.github.com/yas375/6e271111ab382588a0f7b5bba21bbe3b). I haven't inspected the json, but I did check that it has the files I don't see at coveralls.io at all and it passed json lint.

Finally I've tried to use HEAD version of slather

> gem 'slather', git: 'https://github.com/SlatherOrg/slather.git', ref: '79d7652'

And now it [works](https://coveralls.io/builds/6001834)! Probably one of [these](https://github.com/SlatherOrg/slather/compare/v2.1.0...79d7652) commits has fixed it. Not sure which one :)
